### ISSUE TITLE
Parse LUIS builtin.datetime.date in C#

### ISSUE THREAD

**Issue by Botship** [Microsoft/BotBuilder#772]
How to parse LuisResult entities resolution to get actual date?  There seems Node/core/lib/dialogs/EntityRecognizer.js  in JS/Node. Is there any similar method for C#?

e.g. {[date, XXXX-08-22]} to 2016-08-22

**Comment by chrimc62** [comment]
There is not anything built-in.  If you end up developing one we would love to include it.

**Comment by willportnoy** [comment]
It looks like we use Chronic in the alarm bot sample:

https://github.com/Microsoft/BotBuilder/blob/master/CSharp/Samples/SimpleAlarmBot/SimpleAlarmDialog.cs#L105

**Comment by Botship** [comment]
Yes, I saw that but its a bit limited and does not parse all LuisResults e.g. {[date, XXXX-08-22]} , Thanks.

**Comment by kienct89** [comment]
I think Chronic can't parse duration too

**Comment by garypretty** [comment]
@Botship @willportnoy @chrimc62  I am going to look at building an EntityRecognizer for C# dev, so will  let you know how I get on.

**Comment by garypretty** [comment]
@willportnoy @chrimc62 @Botship Hey guys, I want to make sure anything I build here is helpful, so couple of quick questions that would really help me out here.

1. If LUIS passes back a partial date, such as XXXX-WXX-4, would you think it is likely that I should return the date of the next Thursday?  My worry here is that the user may have been referring to "last thursday" for example.

2. If LUIS returns a partial date that is identifying a week, such as XXXX-W50-X, should I pass back some form of duration. i.e. start and end of the next week 50?

I guess I would just be interested to know if you have any other thoughts on what would be useful for a LUIS entity helper in C#.

Thanks

**Comment by garypretty** [comment]
@chrimc62 @willportnoy @Botship Ok, I decided to mirror the functionality available in Node for now and expand later on.  I have not done any testing yet so more changes needed, but thought you might like to see how I am getting on. The C# EntityRecognizer is on GitHub at https://github.com/garypretty/botframework/blob/master/BotFrameworkHelpers/EntityRecognizer.cs 

Once completed I am happy to make this available for inclusion in the framework if you guys want it. Would be great to have the functionality available for C# devs.

Let me know any thoughts, and as I say, I haven't tested this code yet :)

**Comment by garypretty** [comment]
@msft-shahins sorry, should have mentioned your good self as well as it is assigned to you. Please see above for something I am working on with this. Would love to help deliver this and get something in for you.

**Comment by willportnoy** [comment]
It looks like useful code @garypretty.  I have some code fragments to parse the LUIS datetime resolution (using a regex) for a sample I'm working on - I'm trying to get more clarity on the extent of the LUIS datetime resolution language.

**Comment by garypretty** [comment]
@willportnoy sounds great, maybe we can roll in the Regex you have into this Utility class then and put a combination of the two in. I did a ton of queries to LUIS to try and find all the combinations for the LUIS resolution language - the one big omission at the moment is duration, which I am planning on adding (this could do with adding to Node as well). Can we pick this up offline maybe?  I'll add you on Skype and we can chat at some point if that's ok?

**Comment by msft-shahins** [comment]
@garypretty the utility class looks useful and definitely would be good to have it in the SDK. As @willportnoy mentioned we are waiting on some clarity for different formats returned by LUIS for datatime resolution. It would be a good idea if you can integrate your utility class with the builder and send a pull request when you think that it is ready. Also it would be good to have documentation and unit tests for the utility class too.

**Comment by garypretty** [comment]
@msft-shahins Not a problem. When you and @willportnoy get any additional info it would be great if you could share and I can update if I am still in progress and make sure I have the bases covered. Cheers!

**Comment by garypretty** [comment]
Hey. @msft-shahins @willportnoy Hope you are both good. I have now got the EntityRecognizer for C# to a point where it is ready. I have added appropriate XML docs for DoxyGen and I have added a bunch of unit tests as well.  I have tried pushing my local branch up to GitHub so that I can submit the pull request, but I don't seem to have access. Is this something you can provide please?

Is there any specific naming convention you want me to use for the branch name. locally I have called it 'gpretty-csharp-entity-recognizer'

There is more functionality I would like to add at some stage for additional types of entities, but right now it handles the same types as the utility in Node so probably a good time to get an initial version in.

Thanks and speak soon.

**Comment by msft-shahins** [comment]
@garypretty you should be able to send pull request by following standard github process [here](https://help.github.com/articles/using-pull-requests/). Basically, you will fork the repository to your personal account, push your changes to your upstream, and send the pull request against the master that you have forked your repo from.

**Comment by garypretty** [comment]
@msft-shahins @willportnoy hi guys.  Pull request submitted as discussed (thanks for pointing me to the instructions, all pretty straight forward). I updated a bunch of the code to use RegEx for recognizing the LUIS date / time entities, so think we have most, if not all, of the user cases covered. Looking forward to your feedback. Thanks!

**Comment by willportnoy** [comment]
Just to update the thread, @garypretty and I have been continuing this conversation, working to define some shared interfaces for parsing LUIS entities to CLR types e.g.

https://github.com/Microsoft/BotBuilder/blob/develop/CSharp/Library/Luis/EntityToType.cs
https://github.com/Microsoft/BotBuilder/blob/develop/CSharp/Tests/Microsoft.Bot.Builder.Tests/LuisResolutionTests.cs
https://github.com/Microsoft/BotBuilder/blob/develop/CSharp/Library/Luis/Resolution.cs

**Comment by willportnoy** [comment]
I'm going to close this issue for now, but feel free to open new, specific issues for further discussion as we iterate and improve the implementation.

**Comment by anumeshyerigeri** [comment]
builtin.datetime.date is not working while taking date for upcoming month(question) it is showing "FUTURE_REF" in builtin.datetime.date  instead taking as next month last monday it is taking as august month for upcoming month but now it is "FUTURE_REF" why it is changing. please keep it constant

For example: {
      "entity": "upcoming month holidays",
      "type": "builtin.datetime.date",
      "startIndex": 0,
      "endIndex": 22,
      "resolution": {
        "date": "FUTURE_REF"
      }

**Comment by JasonSowers** [comment]
@anumeshyerigeri can you please open a new issue for this
